arch/arm/core/aarch32/mpu/arm_mpu.c

```diff
@@ -326,11 +326,15 @@ int z_arm_mpu_init(void)
 	arm_core_mpu_disable();
 
 #if defined(CONFIG_NOCACHE_MEMORY)
-	/* Clean and invalidate data cache if
+	/* Clean and/or invalidate data cache if
 	 * that was not already done at boot
 	 */
 #if !defined(CONFIG_INIT_ARCH_HW_AT_BOOT)
-	SCB_CleanInvalidateDCache();
+	if (SCB->CCR & SCB_CCR_DC_Msk) {
+		SCB_CleanInvalidateDCache();
+	} else {
+		SCB_InvalidateDCache();
```

Comment:
I believe the `else` branch here is not required. If D-cache is turned off it will be invalidated when it's enabled. The original call to `SCB_CleanInvalidateDCache` was introduced in #27986 to work around an issue on a SoC that boots with D-cache enabled.

Reply:
You are correct, here I'll fix that.

However note that in the `CONFIG_INIT_ARCH_HW_AT_BOOT` case, we do need to call `SCB_InvalidateDCache()` in the else branch, as this is supposed to bring the CPU to a known state. During my testing, I observed that when the cache is disabled, calling `SCB_DisableDCache()` exhibits exactly the same issue than when calling `SCB_CleanInvalidateDCache()`. This is because  `SCB_DisableDCache()` starts by cleaning and invalidating the cache before disabling it. Therefore by calling `SCB_InvalidateDCache()`, we make sure that there won't be any failure later if `SCB_DisableDCache()` is called while the cache is already disabled. For me it looks like a bug in the CMSIS code, but it's not something we can change anyway.

In addition from what I understand, CONFIG_INIT_ARCH_HW_AT_BOOT is supposed to bring the CPU to a known state, so it makes sense to bring the cache to a given state even if it is disabled. The same way that this features clears all the MPU regions and not simply disables the MPU.

Reply:
Sounds good. Thanks for the fix!

Reply:
> In addition from what I understand, CONFIG_INIT_ARCH_HW_AT_BOOT is supposed to bring the CPU to a known state, so it makes sense to bring the cache to a given state even if it is disabled. The same way that this features clears all the MPU regions and not simply disables the MPU.

That is correct - that's why INIT_ARCH_HW_AT_BOOT was introduced.